### PR TITLE
add Git tests for our current pull behaviour

### PR DESCRIPTION
**Pull request by shiftkey**
## Overview

This PR adds some scaffolding and tests for the `pull` Git operation, to provide some coverage before I dive in with rebasing (which will need similar scaffolding).

[**Crafting Git Tests docs**](https://github.com/desktop/desktop/blob/add-some-tests-for-pull/docs/technical/crafting-git-tests.md)

## Description

This PR came from the need to be able to programatically build up repositories to help with testing, instead of unpacking from an archive.

Why go down this path?

 - Archives of repositories are great for capturing regressions, but are essentially a black box of Git state
 - Updating archives over time is a pain (and also really hard to diff).
 - Building up the repository programatically in a test means we can adapt tests as our requirements and understanding of flows change.

An example of this process is the `app/test/helpers/repository-builder-pull-test.ts` function, which is responsible for setting up a repository that we use in the `pull` tests.

I've kept the scaffolding simple for now, but some examples of functions:

 - `makeCommit` - pass in a collection of files which are written to disk, staged and committed
 - `switchTo` - a quick way to ensure you're on the right branch (creates the branch if not found)

The rest of the changes were about testing how our `pull` implementation responds in these scenarios (with different configuration set):

 - ahead only (local commits not on remote)
 - behind only (remote commits not on current branch)
 - ahead and behind (commits both on remote and local that need to be reconciled)

TODO: 

 - [x] review merge conflict tests and see if they are a good candidate for the current API
 - [x] write some docs for scaffolding Git repository tests
 - [x] refine the API so it's good enough for a v1 ship

## Release notes

Notes: no-notes

### CODE REVIEW

**Comment by shiftkey** [comment]
I've marked this as `ready-for-review` despite one open task because I'd love some input on [the docs](https://github.com/desktop/desktop/blob/add-some-tests-for-pull/docs/technical/crafting-git-tests.md) about crafting Git tests (which includes the new and old approaches) as well as feedback on the current scaffolding APIs - I've mostly focused on what I think was needed for the merge conflicts and upcoming rebase flows, so I've kept it narrow for now.

If there are additional patterns you'd like to propose, suggest them in this PR and we can extract tasks and figure out when to address them in the future.

File: app/test/unit/git/pull/ahead-and-behind-test.ts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+    describe('with pull.ff=only set in config', () => {
+      beforeEach(async () => {
+        await GitProcess.exec(
+          ['config', '--local', 'pull.ff', 'only'],
+          repository.path
+        )
+      })
+
+      it(`throws an error as the user blocks merge commits on pull`, async () => {
```
Comment (outofambit) [inline app/test/unit/git/pull/ahead-and-behind-test.ts]: unused async
```suggestion
      it(`throws an error as the user blocks merge commits on pull`, () => {
```

File: app/test/unit/git/pull/only-ahead-test.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        await pull(repository, null, origin)
+
+        newTip = await getTipOrError(repository)
+      })
+
+      it('does not create new commit', async () => {
+        expect(newTip.sha).toBe(previousTip.sha)
+      })
+
+      it('is different to tracking branch', async () => {
```
Comment (outofambit) [inline app/test/unit/git/pull/only-ahead-test.ts]:
```suggestion
      it('is different from tracking branch', async () => {
```

**Comment by shiftkey** [comment]
Merging this in to unblock #6698

**Pull request merged**